File: 1026-Maximum-Difference-Between-Node-and-Ancestor/1026-Maximum-Difference-Between-Node-and-Ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int find(TreeNode* root, int maxN, int minN){
        //base condition for leaf node i.e end of path
        if(root== NULL) return abs(maxN - minN);
        //updating the maxium and minimum value
        maxN = max(root->val, maxN);
        minN = min(root->val, minN);
        //for left traversal
        int left = find(root->left, maxN, minN);
        //for right traversal
        int right = find(root->right, maxN, minN);
        //returning the max value from different path
        return max(left, right);
    }
    int maxAncestorDiff(TreeNode* root) {
        if(root == NULL) return 0;
        return find(root, root->val, root->val);
    }
};

File: 1026-Maximum-Difference-Between-Node-and-Ancestor/1026-Maximum-Difference-Between-Node-and-Ancestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def find(self, root, maxN, minN):
        if not root:
            return abs(maxN - minN)
        maxN = max(root.val, maxN)
        minN = min(root.val, minN)
        left = self.find(root.left, maxN, minN)
        right = self.find(root.right, maxN, minN)
        return max(left, right)
    def maxAncestorDiff(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        return self.find(root, root.val, root.val)
